### PR TITLE
[core] Minor cleanup for object recovery manager

### PR DESCRIPTION
**Pull request by dentiny**
As usual, cleanup some C++ usage for object reference manager.

### CODE REVIEW

File: src/ray/core_worker/object_recovery_manager.cc [hunk 1]
```diff
@@ -55,10 +55,12 @@ bool ObjectRecoveryManager::RecoverObject(const ObjectID &object_id) {
...
-    in_memory_store_->GetAsync(
+    in_memory_store_.GetAsync(
         object_id, [this, object_id](std::shared_ptr<RayObject> obj) {
-          absl::MutexLock lock(&mu_);
-          RAY_CHECK(objects_pending_recovery_.erase(object_id)) << object_id;
+          {
+            absl::MutexLock lock(&mu_);
+            RAY_CHECK(objects_pending_recovery_.erase(object_id)) << object_id;
+          }
           RAY_LOG(INFO).WithField(object_id) << "Recovery complete for object";
```
Comment (dentiny) [inline src/ray/core_worker/object_recovery_manager.cc:64]: `RAY_LOG` could performance IO operation, shouldn't be placed within critical section.

File: src/ray/core_worker/object_recovery_manager.h [hunk 2]
```diff
@@ -46,26 +43,25 @@ class ObjectRecoveryManager {
...
+        client_factory_(std::move(client_factory)),
         local_object_pinning_client_(local_object_pinning_client),
-        object_lookup_(object_lookup),
+        object_lookup_(std::move(object_lookup)),
         in_memory_store_(in_memory_store),
-        recovery_failure_callback_(recovery_failure_callback) {}
+        recovery_failure_callback_(std::move(recovery_failure_callback)) {}
 
   /// Recover an object that was stored in plasma. This will only succeed for
-  /// objects that are lost from memory and that this process owns (returns
```
Comment (dentiny) [inline src/ray/core_worker/object_recovery_manager.h:63]: I find it hard to understand what does it mean by "that this process owns".

File: src/ray/core_worker/object_recovery_manager.cc [hunk 3]
```diff
@@ -86,9 +88,10 @@ void ObjectRecoveryManager::PinOrReconstructObject(
     const ObjectID &object_id, const std::vector<rpc::Address> &locations) {
   RAY_LOG(DEBUG).WithField(object_id)
       << "Lost object has " << locations.size() << " locations";
+  // The object to recovery is stored elsewhere, attempt to pin the object.
```
Comment (jjyao) [inline src/ray/core_worker/object_recovery_manager.cc:91]:
```suggestion
  // The object to recovery has secondary copies, pin one copy to promote it to primary one.
```

File: src/ray/core_worker/object_recovery_manager.cc [hunk 4]
```diff
@@ -132,21 +135,22 @@ void ObjectRecoveryManager::PinExistingObjectCopy(
...
+                           RAY_CHECK(in_memory_store_.Put(
                                RayObject(rpc::ErrorType::OBJECT_IN_PLASMA), object_id));
-                           reference_counter_->UpdateObjectPinnedAtRaylet(object_id,
-                                                                          node_id);
+                           reference_counter_.UpdateObjectPinnedAtRaylet(object_id,
+                                                                         node_id);
                          } else {
                            RAY_LOG(INFO).WithField(object_id)
-                               << "Error pinning new copy of lost object, trying again";
+                               << "Error pinning new copy of lost object since " << status
```
Comment (jjyao) [inline src/ray/core_worker/object_recovery_manager.cc:144]:
```suggestion
                               << "Error pinning secondary copy of lost object due to " << status
```

**Review by jjyao — COMMENTED** [review]
LG

File: src/ray/core_worker/object_recovery_manager.cc (hunk 3 above)
Comment (dentiny) [reply src/ray/core_worker/object_recovery_manager.cc:91]: Updated.

File: src/ray/core_worker/object_recovery_manager.cc (hunk 4 above)
Comment (dentiny) [reply src/ray/core_worker/object_recovery_manager.cc:144]: Updated.

**Review by jjyao — APPROVED** [review]

**Pull request merged**